File: day-02/solution.ts
import { readLines } from "../commons/io.ts";

/*
 * Sample Input 
Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green
 */

function main() {
  const lines = readLines("day-02/input.txt");
  let gameIdSum = 0;
  let minPowerSum = 0;
  lines.forEach((line) => {
    const gameInfo = gameParser(line);
    let allGamesValid = true;
    minPowerSum += findPower(findMinimumSet(gameInfo));
    for (const game of gameInfo.games) {
      if (!gameValidator(game)) {
        allGamesValid = false;
        return;
      }
    }
    if (allGamesValid) {
      gameIdSum += gameInfo.gameID;
    }

  });

  console.log("Game ID Sum: ", gameIdSum);
  console.log("Minimum Power Sum: ", minPowerSum);
}

const RED_MAX = 12;
const GREEN_MAX = 13;
const BLUE_MAX = 14;

type Color = "red" | "green" | "blue";

class Game {
  red = 0;
  green = 0;
  blue = 0;
}

class GameInfo {
  gameID = 0;
  games: Game[] = [];
}

function gameParser(line: string): GameInfo {
  const [gameIDstr, gameDatas] = line.split(":");
  const gameID = gameIDstr.split(" ")[1];
  const games: Game[] = [];
  const gameStrings = gameDatas.split(";");
  gameStrings.forEach((gameString) => {
    const colorData = gameString.split(",");
    const game = new Game();
    colorData.forEach((colorData) => {
      const [num, color] = colorData.trim().split(" ");
      game[color as Color] = +num;
    });
    games.push(game);
  });
  return {
    gameID: +gameID,
    games,
  };
}

function gameValidator(game: Game) {
  const { red, green, blue } = game;
  if (red > RED_MAX || green > GREEN_MAX || blue > BLUE_MAX) {
    return false;
  }
  return true;
}

function findMinimumSet(gameInfo: GameInfo) {
  const { games } = gameInfo;
  const minimumSet: Game = { red: 0, green: 0, blue: 0}
  for (const game of games) {
    if (game.red > minimumSet.red) {
      minimumSet.red = game.red;
    }
    if (game.green > minimumSet.green) {
      minimumSet.green = game.green;
    }
    if (game.blue > minimumSet.blue) {
      minimumSet.blue = game.blue;
    }
  }
  return minimumSet;
}

function findPower(game: Game) {
  return game.red * game.green * game.blue;
}

main();

File: day-02/solution.py
# pylint: disable=missing-module-docstring,missing-function-docstring,import-error,wrong-import-position,too-few-public-methods
import sys
import os

sys.path.append(os.path.join(os.path.dirname(__file__), ".."))
from typing import List, Protocol
from commons.io import readLines

RED_MAX = 12
GREEN_MAX = 13
BLUE_MAX = 14


class GameInfo(Protocol):
    """
    Represents information about a game.
    """
    game_id: int
    games: List[dict]


class Game:
    """
    Represents a game.
    """
    def __init__(self, red = 0, green = 0, blue = 0):
        self.red = red
        self.green = green
        self.blue = blue


def game_parser(line) -> GameInfo:
    game_id_str, game_datas = line.split(":")
    game_id = game_id_str.split(" ")[1]
    games = []
    game_strings = game_datas.split(";")
    for game_string in game_strings:
        color_datas = game_string.split(",")
        game = Game()
        for color_data in color_datas:
            num, color = color_data.strip().split(" ")
            if color == "red":
                game.red = int(num)
            elif color == "green":
                game.green = int(num)
            elif color == "blue":
                game.blue = int(num)
        games.append(game)
    return {
        "game_id": int(game_id),
        "games": games,
    }


def game_validator(game):
    red = game.red
    green = game.green
    blue = game.blue
    if red > RED_MAX or green > GREEN_MAX or blue > BLUE_MAX:
        return False
    return True


def find_minimum_set(game_info) -> Game:
    games = game_info["games"]
    minimum_set = Game()
    for game in games:
        if game.red > minimum_set.red:
            minimum_set.red = game.red
        if game.green > minimum_set.green:
            minimum_set.green = game.green
        if game.blue > minimum_set.blue:
            minimum_set.blue = game.blue
    return minimum_set


def find_power(game):
    return game.red * game.green * game.blue


def main():
    lines: List[str] = readLines("day-02/input.txt")
    game_id_sum = 0
    min_power_sum = 0
    for line in lines:
        game_info = game_parser(line)
        all_games_valid = True
        min_power_sum += find_power(find_minimum_set(game_info))
        for game in game_info["games"]:
            if not game_validator(game):
                all_games_valid = False
                break
        if all_games_valid:
            game_id_sum += game_info["game_id"]

    print("Game ID Sum: ", game_id_sum)
    print("Minimum Power Sum: ", min_power_sum)


main()
